src/csharp/Grpc.Tools/GeneratorServices.cs

```diff
@@ -109,6 +108,26 @@ string LowerUnderscoreToUpperCamel(string str)
             }
             return result.ToString();
         }
+
+        // This is how the protoc codegen constructs its output filename.
+        // See protobuf/compiler/csharp/csharp_helpers.cc:356.
```

Comment:
for comparison,  the logic is here:
https://github.com/protocolbuffers/protobuf/blob/3a3956e8a258784461270961c6577341356bce52/src/google/protobuf/compiler/csharp/csharp_helpers.cc#L137  with `cap_next_letter=true` and `preserve_period=false`

Reply:
Done.